File: bench/metal/matmul_2dev_distributed.cpp
// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
// SPDX-License-Identifier: Apache-2.0
//
// Distributed matmul benchmark on 2-device N300 (1x2 mesh)
// Both devices run the same matmul in parallel (data parallel)
// This measures the overhead of mesh device vs single device

#include <ttnn/device.hpp>
#include <ttnn/types.hpp>
#include <ttnn/tensor/shape/shape.hpp>
#include <ttnn/operations/core/core.hpp>
#include <ttnn/operations/creation.hpp>
#include <ttnn/operations/matmul/matmul.hpp>
#include <ttnn/distributed/api.hpp>
#include <ttnn/distributed/types.hpp>
#include <tt-metalium/distributed.hpp>
#include <tt-metalium/mesh_config.hpp>
#include <tt-metalium/dispatch_core_common.hpp>

#include <array>
#include <chrono>
#include <vector>
#include <algorithm>
#include <numeric>
#include <cmath>
#include <set>

using namespace ttnn;
using namespace tt::tt_metal::distributed;

constexpr int N_WARMUP = 5;
constexpr int N_ITER = 20;

struct BenchResult {
    uint32_t size;
    double mean_us;
    double std_us;
    double min_us;
    double max_us;
    double tflops;
};

BenchResult benchmark_mesh_matmul(MeshDevice& mesh_device, uint32_t size, int num_devices) {
    std::array<uint32_t, 2> shape = {size, size};

    // Create tensors on mesh device (replicated to all devices)
    auto a = ones(Shape(shape), DataType::BFLOAT16, TILE_LAYOUT, mesh_device);
    auto b = ones(Shape(shape), DataType::BFLOAT16, TILE_LAYOUT, mesh_device);

    MeshCommandQueue& cq = mesh_device.mesh_command_queue();

    // Warmup
    for (int i = 0; i < N_WARMUP; i++) {
        auto c = ttnn::matmul(a, b);
        Finish(cq);
    }

    // Timed runs
    std::vector<double> times_us(N_ITER);

    for (int i = 0; i < N_ITER; i++) {
        auto start = std::chrono::high_resolution_clock::now();
        auto c = ttnn::matmul(a, b);
        Finish(cq);
        auto end = std::chrono::high_resolution_clock::now();
        times_us[i] = std::chrono::duration<double, std::micro>(end - start).count();
    }

    // Stats
    double sum = std::accumulate(times_us.begin(), times_us.end(), 0.0);
    double mean = sum / N_ITER;

    double sq_sum = 0.0;
    for (double t : times_us) {
        sq_sum += (t - mean) * (t - mean);
    }
    double std_dev = std::sqrt(sq_sum / N_ITER);

    std::sort(times_us.begin(), times_us.end());

    // TFLOP/s: num_devices * 2 * N^3 / time (each device does full matmul)
    double flops = static_cast<double>(num_devices) * 2.0 * std::pow(size, 3);
    double tflops = flops / (mean * 1e-6) / 1e12;

    return BenchResult{
        .size = size,
        .mean_us = mean,
        .std_us = std_dev,
        .min_us = times_us.front(),
        .max_us = times_us.back(),
        .tflops = tflops,
    };
}

std::vector<uint32_t> generate_test_sizes() {
    std::set<uint32_t> sizes;

    // Powers of 2
    for (int exp = 5; exp <= 12; exp++) {  // 32 to 4096
        sizes.insert(1 << exp);
    }

    // Fine sweep around inflection points
    struct Region { uint32_t start, end, step; };
    std::vector<Region> regions = {
        {192, 384, 32},
        {384, 640, 32},
        {896, 1152, 32},
        {1920, 2176, 32},
        {3840, 4096, 64},
    };

    for (const auto& r : regions) {
        for (uint32_t s = r.start; s <= r.end; s += r.step) {
            if (s % 32 == 0) {
                sizes.insert(s);
            }
        }
    }

    return std::vector<uint32_t>(sizes.begin(), sizes.end());
}

int main() {
    // Open 1x2 mesh device (2 devices)
    auto mesh_device = ttnn::distributed::open_mesh_device(
        MeshShape(1, 2),
        DEFAULT_L1_SMALL_SIZE,
        DEFAULT_TRACE_REGION_SIZE,
        1,  // num_command_queues
        tt::tt_metal::DispatchCoreConfig{}
    );

    int num_devices = mesh_device->num_devices();
    auto sizes = generate_test_sizes();

    fmt::print("# Mesh matmul benchmark on 1x2 MeshDevice (N300)\n");
    fmt::print("# {} devices running matmul in parallel (data parallel)\n", num_devices);
    fmt::print("size,mean_us,std_us,min_us,max_us,tflops\n");

    for (uint32_t size : sizes) {
        try {
            auto result = benchmark_mesh_matmul(*mesh_device, size, num_devices);
            fmt::print("{},{:.2f},{:.2f},{:.2f},{:.2f},{:.3f}\n",
                       result.size, result.mean_us, result.std_us,
                       result.min_us, result.max_us, result.tflops);
        } catch (const std::exception& e) {
            fmt::print(stderr, "# Error at size {}: {}\n", size, e.what());
        }
    }

    ttnn::distributed::close_mesh_device(mesh_device);
    return 0;
}

File: scripts/matmul_2dev_distributed.py
#!/usr/bin/env python3
"""
Distributed matmul benchmark on 2-device N300 (1x2 mesh).

Pattern: Column-parallel tensor parallelism
- Replicate A to both devices
- Shard B by columns (dim=1) across devices
- Each device computes half the output columns
- all_gather to combine results

This actually distributes work across both devices, unlike replicating
both tensors which runs redundant computation on each device.
"""

import os
import sys
import time
import numpy as np

# Setup TT-Metal environment
if "TT_METAL_HOME" not in os.environ:
    os.environ["TT_METAL_HOME"] = os.path.expanduser("~/tt-metal")
if "ARCH_NAME" not in os.environ:
    os.environ["ARCH_NAME"] = "wormhole_b0"

import torch
import ttnn

N_WARMUP = 5
N_ITER = 20


def benchmark_distributed_matmul(mesh_device, size: int, include_allgather: bool = True) -> dict:
    """
    Benchmark distributed matmul with column-parallel sharding.

    Args:
        mesh_device: 1x2 MeshDevice
        size: Matrix dimension (must be divisible by 32 for tiles and 2 for sharding)
        include_allgather: Whether to include all_gather in timing (True for e2e)

    Returns:
        dict with timing stats and throughput
    """
    # Create input tensors
    a = torch.randn(size, size, dtype=torch.bfloat16)
    b = torch.randn(size, size, dtype=torch.bfloat16)

    # Replicate A to both devices
    a_tt = ttnn.from_torch(
        a,
        layout=ttnn.TILE_LAYOUT,
        device=mesh_device,
        dtype=ttnn.bfloat16,
        mesh_mapper=ttnn.ReplicateTensorToMesh(mesh_device),
    )

    # Shard B by columns (dim=1) - each device gets half the columns
    b_tt = ttnn.from_torch(
        b,
        layout=ttnn.TILE_LAYOUT,
        device=mesh_device,
        dtype=ttnn.bfloat16,
        mesh_mapper=ttnn.ShardTensorToMesh(mesh_device, dim=1),
    )

    # Warmup
    for _ in range(N_WARMUP):
        c_sharded = ttnn.matmul(a_tt, b_tt)
        if include_allgather:
            c_full = ttnn.all_gather(
                c_sharded,
                dim=1,
                num_links=1,
                topology=ttnn.Topology.Linear,
            )
        ttnn.synchronize_device(mesh_device)

    # Timed runs
    times = []
    for _ in range(N_ITER):
        t0 = time.perf_counter()
        c_sharded = ttnn.matmul(a_tt, b_tt)
        if include_allgather:
            c_full = ttnn.all_gather(
                c_sharded,
                dim=1,
                num_links=1,
                topology=ttnn.Topology.Linear,
            )
        ttnn.synchronize_device(mesh_device)
        t1 = time.perf_counter()
        times.append((t1 - t0) * 1e6)  # microseconds

    times = np.array(times)

    # FLOPS: 2 * N^3 for the full matmul (split across 2 devices)
    # Since we're doing the equivalent of a full N×N × N×N matmul
    flops = 2.0 * (size ** 3)
    tflops = flops / (times.mean() * 1e-6) / 1e12

    return {
        "size": size,
        "mean_us": times.mean(),
        "std_us": times.std(),
        "min_us": times.min(),
        "max_us": times.max(),
        "tflops": tflops,
    }


def benchmark_single_device_matmul(mesh_device, size: int) -> dict:
    """
    Benchmark single-device matmul for comparison.
    Uses only device 0 in the mesh.
    """
    a = torch.randn(size, size, dtype=torch.bfloat16)
    b = torch.randn(size, size, dtype=torch.bfloat16)

    # Get first device from mesh
    device = mesh_device.get_device(0)

    a_tt = ttnn.from_torch(
        a,
        layout=ttnn.TILE_LAYOUT,
        device=device,
        dtype=ttnn.bfloat16,
    )
    b_tt = ttnn.from_torch(
        b,
        layout=ttnn.TILE_LAYOUT,
        device=device,
        dtype=ttnn.bfloat16,
    )

    # Warmup
    for _ in range(N_WARMUP):
        c = ttnn.matmul(a_tt, b_tt)
        ttnn.synchronize_device(device)

    # Timed runs
    times = []
    for _ in range(N_ITER):
        t0 = time.perf_counter()
        c = ttnn.matmul(a_tt, b_tt)
        ttnn.synchronize_device(device)
        t1 = time.perf_counter()
        times.append((t1 - t0) * 1e6)

    times = np.array(times)
    flops = 2.0 * (size ** 3)
    tflops = flops / (times.mean() * 1e-6) / 1e12

    return {
        "size": size,
        "mean_us": times.mean(),
        "std_us": times.std(),
        "min_us": times.min(),
        "max_us": times.max(),
        "tflops": tflops,
    }


def generate_test_sizes():
    """Generate test sizes (must be divisible by 32 for tiles and 64 for 2-way sharding)."""
    sizes = []

    # Powers of 2 (all are divisible by 64)
    for exp in range(6, 13):  # 64 to 4096
        sizes.append(2 ** exp)

    # Additional sizes around inflection points (divisible by 64)
    additional = [
        192, 256, 320, 384, 448, 512, 576, 640,  # Early region
        896, 960, 1024, 1088, 1152,               # Core saturation
        1920, 1984, 2048, 2112, 2176,             # Memory boundary
        3840, 3904, 3968, 4032, 4096,             # Large size
    ]

    for s in additional:
        if s not in sizes and s % 64 == 0:
            sizes.append(s)

    return sorted(set(sizes))


def main():
    import argparse
    parser = argparse.ArgumentParser(description="Distributed matmul benchmark on N300")
    parser.add_argument("--mode", choices=["distributed", "single", "both"], default="both",
                        help="Benchmark mode: distributed, single-device, or both")
    parser.add_argument("--no-allgather", action="store_true",
                        help="Skip all_gather in distributed mode (just measure matmul)")
    parser.add_argument("--sizes", type=str, default=None,
                        help="Comma-separated list of sizes to test")
    args = parser.parse_args()

    if args.sizes:
        sizes = [int(s) for s in args.sizes.split(",")]
    else:
        sizes = generate_test_sizes()

    print(f"# Distributed matmul benchmark on 1x2 MeshDevice (N300)", file=sys.stderr)
    print(f"# Mode: {args.mode}, include_allgather: {not args.no_allgather}", file=sys.stderr)
    print(f"# Testing {len(sizes)} sizes", file=sys.stderr)

    # Open 1x2 mesh device
    mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 2))

    try:
        if args.mode in ["distributed", "both"]:
            print("\n# Distributed (column-parallel) matmul")
            print("mode,size,mean_us,std_us,min_us,max_us,tflops")

            for size in sizes:
                try:
                    result = benchmark_distributed_matmul(
                        mesh_device, size,
                        include_allgather=not args.no_allgather
                    )
                    print(f"distributed,{result['size']},{result['mean_us']:.2f},"
                          f"{result['std_us']:.2f},{result['min_us']:.2f},"
                          f"{result['max_us']:.2f},{result['tflops']:.3f}")
                    sys.stdout.flush()
                except Exception as e:
                    print(f"# Error at distributed size {size}: {e}", file=sys.stderr)
                    import traceback
                    traceback.print_exc()

        if args.mode in ["single", "both"]:
            print("\n# Single-device matmul (for comparison)")
            print("mode,size,mean_us,std_us,min_us,max_us,tflops")

            for size in sizes:
                try:
                    result = benchmark_single_device_matmul(mesh_device, size)
                    print(f"single,{result['size']},{result['mean_us']:.2f},"
                          f"{result['std_us']:.2f},{result['min_us']:.2f},"
                          f"{result['max_us']:.2f},{result['tflops']:.3f}")
                    sys.stdout.flush()
                except Exception as e:
                    print(f"# Error at single size {size}: {e}", file=sys.stderr)

    finally:
        ttnn.close_mesh_device(mesh_device)


if __name__ == "__main__":
    main()
